### PR TITLE
added PCA

### PR DESCRIPTION
**Pull request by valinurovdenis**
Signed-off-by: valinurovdenis <denis.valinurow@yandex.ru>

### CODE REVIEW

File: NeoML/include/NeoML/TraditionalML/PCA.h [hunk 1]
```diff
@@ -0,0 +1,92 @@
...
+	~CPca() {};
+
+	// Chooses `Components` greatest singular values and
+	// selects the corresponding principal axis as the final components
+	void Train( const CFloatMatrixDesc& data );
+	// Train + transform the data into shape ( samples x components )
+	CSparseFloatMatrixDesc Transform( const CFloatMatrixDesc& data );
+
+	// Singular values corresponding to the selected principal axis
+	void GetSingularValues( CArray<float>& values ) const { singularValues.CopyTo( values ); }
```
Comment (SAngeliuk) [inline NeoML/include/NeoML/TraditionalML/PCA.h]: const CArray<float>& GetSingularValues() const; ?

File: NeoML/include/NeoML/TraditionalML/PCA.h [hunk 2]
```diff
@@ -0,0 +1,92 @@
...
+	void GetExplainedVarianceRatio( CArray<float>& values ) const { explainedVarianceRatio.CopyTo( values ); }
+	// Mean of singular values not corresponding to the selected principal axis
+	float GetNoiseVariance() const { return noiseVariance; }
+	// Selected number of principal axis
+	int GetComponentsNum() const { return components; }
+	// Matrix ( components x features ) with rows corresponding to the selected principal axis 
+	CFloatMatrixDesc GetComponents() { return componentsMatrix.GetDesc(); }
+
+protected:
+	CArray<float> singularValues;
```
Comment (SAngeliuk) [inline NeoML/include/NeoML/TraditionalML/PCA.h:104]: -> private

File: NeoML/src/TraditionalML/PCA.cpp [hunk 3]
```diff
@@ -0,0 +1,199 @@
...
+#include <common.h>
+#pragma hdrstop
+
+#include <NeoML/Dnn/DnnBlob.h>
+#include <NeoML/TraditionalML/PCA.h>
+#include <memory>
+
+using namespace NeoML;
+
+CPca::CPca( const CParams& _params ): params( _params )
```
Comment (SAngeliuk) [inline NeoML/src/TraditionalML/PCA.cpp]: codestyle!

File: NeoMathEngine/include/NeoMathEngine/NeoMathEngine.h [hunk 4]
```diff
@@ -539,6 +539,12 @@ class NEOMATHENGINE_API IBlasEngine : public IVectorMathEngine {
 	virtual void MatrixSpreadRows( const CConstIntHandle& sourceHandle, int height, int width,
 		const CIntHandle& resultHandle, int resultHeight, const CConstIntHandle& indexHandle,
 		const CConstIntHandle& fillValue ) = 0;
+
+	// Computes the singular value decomposition of matrix `a` of shape m x n
+	// into `a` = `u` * `s` * `vt`, where `u` of shape m x min(n, m), `vt` of shape min(n, m) x n,
+	// `s` is the diagonal matrix of size min(n, m)
+	virtual void SingularValueDecomposition( const CFloatHandle& a, int n, int m, const CFloatHandle& u, const CFloatHandle& s,
```
Comment (SAngeliuk) [inline NeoMathEngine/include/NeoMathEngine/NeoMathEngine.h:551]: What is superb here?

File: NeoMathEngine/src/GPU/Vulkan/VulkanMathEngineBlas.cpp [hunk 5]
```diff
@@ -1152,6 +1152,12 @@ void CVulkanMathEngine::MultiplyVectorByTransposedLookupVectorAndAddToTable( int
 	ASSERT_EXPR( false );
 }
 
+void CVulkanMathEngine::SingularValueDecomposition( const CFloatHandle& a, int n, int m, const CFloatHandle& u, const CFloatHandle& s,
```
Comment (SAngeliuk) [inline NeoMathEngine/src/GPU/Vulkan/VulkanMathEngineBlas.cpp]: warning for unused variables?

File: NeoML/src/TraditionalML/PCA.cpp [hunk 6]
```diff
@@ -0,0 +1,199 @@
...
+{
+	NeoAssert( ( params.ComponentsType == PCAC_None ) ||
+		( ( params.ComponentsType == PCAC_Int ) && ( params.Components > 0 ) ) ||
+		( ( params.ComponentsType == PCAC_Float ) && ( 0 < params.Components ) && ( params.Components < 1 ) ) );
+}
+
+static CPtr<CDnnBlob> convertToBlob( IMathEngine& mathEngine, const CFloatMatrixDesc& data )
```
Comment (SAngeliuk) [inline NeoML/src/TraditionalML/PCA.cpp:43]: Comment all static functions here!

**Review by NadyaG — DISMISSED** [review]
Checked the header comments.

**Comment by NadyaG** [comment]
@SAngeliuk added the documentation on PCA into "miscellaneous algorithms" both in .md files and to the readthedocs sources. Please have a look at the English version. Should I make a Russian translation for the .md?

**Pull request closed without merging**